File: backend/routes/reservations.js
// ===== reservations.js =====
const express = require('express');
const Joi = require('joi');
const db = require('../config/database');
const { authenticateToken, authorizeRoles } = require('../middleware/auth');

const router = express.Router();

// Validation Schemas
const createReservationSchema = Joi.object({
  property_id: Joi.number().integer().required(),
  check_in_date: Joi.date().required(),
  check_out_date: Joi.date().greater(Joi.ref('check_in_date')).required(),
  number_of_guests: Joi.number().integer().min(1).required()
});

const updateStatusSchema = Joi.object({
  status: Joi.string().valid('confirmed', 'cancelled', 'completed').required(),
  reason: Joi.string().max(500).optional()
});

// =====================================================
// CREATE RESERVATION
// =====================================================
router.post('/', authenticateToken, authorizeRoles('tenant'), async (req, res) => {
  try {
    const { error, value } = createReservationSchema.validate(req.body);
    if (error) return res.status(400).json({ message: error.details[0].message });

    const { property_id, check_in_date, check_out_date, number_of_guests } = value;

    const [properties] = await db.execute(
      'SELECT property_id, max_guests, price_per_night, cleaning_fee, is_active FROM properties WHERE property_id = ?',
      [property_id]
    );

    if (properties.length === 0) {
      return res.status(404).json({ message: 'Property not found' });
    }

    const property = properties[0];
    if (!property.is_active) {
      return res.status(400).json({ message: 'Property is not active' });
    }

    if (number_of_guests > property.max_guests) {
      return res.status(400).json({ message: `Maximum ${property.max_guests} guests allowed` });
    }

    const [availabilityCheck] = await db.execute(
      'SELECT fn_check_availability(?, ?, ?) AS is_available',
      [property_id, check_in_date, check_out_date]
    );

    if (!availabilityCheck[0].is_available) {
      return res.status(400).json({ message: 'Selected dates are not available' });
    }

    const inDate = new Date(check_in_date);
    const outDate = new Date(check_out_date);
    const totalNights = Math.ceil((outDate - inDate) / (1000 * 60 * 60 * 24));
    const subtotal = totalNights * property.price_per_night;
    const cleaningFee = property.cleaning_fee || 0;
    const totalAmount = subtotal + cleaningFee;

    const [result] = await db.execute(`
      INSERT INTO reservations (
        property_id, tenant_id, check_in_date, check_out_date,
        guest_count, total_nights, price_per_night, cleaning_fee, total_amount
      ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)`,
      [property_id, req.user.user_id, check_in_date, check_out_date, number_of_guests, totalNights, property.price_per_night, cleaningFee, totalAmount]
    );

    res.status(201).json({
      message: 'Reservation created successfully',
      reservation_id: result.insertId,
      total_nights: totalNights,
      total_amount: totalAmount
    });

  } catch (error) {
    console.error('Reservation creation error:', error);
    res.status(500).json({ message: 'Failed to create reservation' });
  }
});

// =====================================================
// GET MY RESERVATIONS (alias for compatibility)
// =====================================================
router.get('/my', authenticateToken, async (req, res) => {
  try {
    const { status, limit = 10, offset = 0 } = req.query;

    let query = `
      SELECT r.*, 
             p.title AS property_title, p.address, p.city,
             p.price_per_night, p.cleaning_fee,
             u.first_name AS owner_first_name, u.last_name AS owner_last_name,
             u.email AS owner_email, u.phone AS owner_phone,
             pay.payment_status, pay.payment_date,
             pi.image_url AS property_image
      FROM reservations r
      JOIN properties p ON r.property_id = p.property_id
      JOIN users u ON p.owner_id = u.user_id
      LEFT JOIN payments pay ON r.reservation_id = pay.reservation_id
      LEFT JOIN property_images pi ON p.property_id = pi.property_id AND pi.is_primary = 1
      WHERE r.tenant_id = ?
    `;

    const queryParams = [req.user.user_id];
    if (status) {
      query += ' AND r.reservation_status = ?';
      queryParams.push(status);
    }

    query += ' ORDER BY r.booking_date DESC LIMIT ? OFFSET ?';
    queryParams.push(parseInt(limit), parseInt(offset));

    const [reservations] = await db.execute(query, queryParams);

    res.json({
      reservations,
      pagination: {
        limit: parseInt(limit),
        offset: parseInt(offset),
        hasMore: reservations.length === parseInt(limit)
      }
    });

  } catch (error) {
    console.error('Fetch my reservations error:', error);
    res.status(500).json({ message: 'Failed to fetch reservations' });
  }
});

// =====================================================
// UPDATE RESERVATION STATUS
// =====================================================
router.put('/:reservationId/status', authenticateToken, async (req, res) => {
  try {
    const { error, value } = updateStatusSchema.validate(req.body);
    if (error) return res.status(400).json({ message: error.details[0].message });

    const { status, reason } = value;
    const reservationId = req.params.reservationId;

    const [reservations] = await db.execute(`
      SELECT r.*, p.owner_id 
      FROM reservations r
      JOIN properties p ON r.property_id = p.property_id
      WHERE r.reservation_id = ?`,
      [reservationId]
    );

    if (reservations.length === 0) {
      return res.status(404).json({ message: 'Reservation not found' });
    }

    const reservation = reservations[0];
    const isOwner = reservation.owner_id === req.user.user_id;
    const isTenant = reservation.tenant_id === req.user.user_id;
    const isAdmin = req.user.user_role === 'admin';

    if (status === 'confirmed' && !isOwner && !isAdmin) {
      return res.status(403).json({ message: 'Only the owner can confirm' });
    }
    if (status === 'cancelled' && !isOwner && !isTenant && !isAdmin) {
      return res.status(403).json({ message: 'Not authorized to cancel' });
    }
    if (status === 'completed' && !isOwner && !isAdmin) {
      return res.status(403).json({ message: 'Only the owner can complete' });
    }

    let updateQuery = 'UPDATE reservations SET reservation_status = ?, updated_at = CURRENT_TIMESTAMP';
    const updateParams = [status];

    if (status === 'cancelled' && reason) {
      updateQuery += ', cancellation_reason = ?, cancellation_date = CURRENT_TIMESTAMP';
      updateParams.push(reason);
    }

    updateQuery += ' WHERE reservation_id = ?';
    updateParams.push(reservationId);

    await db.execute(updateQuery, updateParams);

    res.json({ message: `Reservation ${status}`, reservation_id: reservationId });

  } catch (error) {
    console.error('Update reservation status error:', error);
    res.status(500).json({ message: 'Status update failed' });
  }
});

module.exports = router;

File: routes/reservations.py

from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity, get_jwt
from utils.database import execute_query
from datetime import datetime

reservations_bp = Blueprint('reservations', __name__)

@reservations_bp.route('/', methods=['POST'])
@jwt_required()
def create_reservation():
    try:
        claims = get_jwt()
        if claims.get('role') != 'tenant':
            return jsonify({'message': 'Only tenants can make reservations'}), 403
        
        data = request.get_json()
        
        # Gerekli alanları kontrol et
        required_fields = ['propertyId', 'checkIn', 'checkOut', 'guests']
        for field in required_fields:
            if not data.get(field):
                return jsonify({'message': f'{field} is required'}), 400
        
        property_id = data['propertyId']
        check_in = data['checkIn']
        check_out = data['checkOut']
        guests = data['guests']
        user_id = int(get_jwt_identity())
        
        # Mülkün varlığını kontrol et
        property = execute_query(
            "SELECT property_id, max_guests, price_per_night, cleaning_fee, is_active FROM properties WHERE property_id = %s",
            (property_id,), fetch_one=True
        )
        
        if not property:
            return jsonify({'message': 'Property not found'}), 404
        
        if not property['is_active']:
            return jsonify({'message': 'Property is not available for booking'}), 400
        
        if guests > property['max_guests']:
            return jsonify({'message': f'Maximum {property["max_guests"]} guests allowed'}), 400
        
        # Müsaitlik kontrolü
        conflicts = execute_query(
            """SELECT COUNT(*) as conflict_count
               FROM reservations
               WHERE property_id = %s
               AND reservation_status IN ('confirmed', 'pending')
               AND (
                 (check_in_date <= %s AND check_out_date > %s) OR
                 (check_in_date < %s AND check_out_date >= %s) OR
                 (check_in_date >= %s AND check_out_date <= %s)
               )""",
            (property_id, check_in, check_in, check_out, check_out, check_in, check_out),
            fetch_one=True
        )
        
        if conflicts['conflict_count'] > 0:
            return jsonify({'message': 'Selected dates are not available'}), 400
        
        # Fiyat hesaplama
        check_in_date = datetime.fromisoformat(check_in.replace('Z', '+00:00'))
        check_out_date = datetime.fromisoformat(check_out.replace('Z', '+00:00'))
        total_nights = (check_out_date - check_in_date).days
        subtotal = total_nights * property['price_per_night']
        cleaning_fee = property['cleaning_fee'] or 0
        total_amount = subtotal + cleaning_fee
        
        # Rezervasyon oluştur
        result = execute_query(
            """INSERT INTO reservations (
                property_id, tenant_id, check_in_date, check_out_date, 
                guest_count, total_nights, price_per_night, cleaning_fee, total_amount
            ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s)""",
            (property_id, user_id, check_in, check_out, guests, total_nights, 
             property['price_per_night'], cleaning_fee, total_amount)
        )
        
        return jsonify({
            'message': 'Reservation created successfully',
            'reservationId': result['last_insert_id'],
            'totalAmount': total_amount,
            'totalNights': total_nights
        }), 201
        
    except Exception as e:
        return jsonify({'message': 'Failed to create reservation', 'error': str(e)}), 500

@reservations_bp.route('/my-reservations', methods=['GET'])
@jwt_required()
def get_my_reservations():
    try:
        user_id = int(get_jwt_identity())
        status = request.args.get('status')
        limit = int(request.args.get('limit', 10))
        offset = int(request.args.get('offset', 0))
        
        query = """
        SELECT r.*, 
               p.title as property_title, p.address, p.city,
               p.price_per_night, p.cleaning_fee,
               u.first_name as owner_first_name, u.last_name as owner_last_name,
               u.email as owner_email, u.phone as owner_phone
        FROM reservations r
        JOIN properties p ON r.property_id = p.property_id
        JOIN users u ON p.owner_id = u.user_id
        WHERE r.tenant_id = %s
        """
        
        params = [user_id]
        
        if status:
            query += " AND r.reservation_status = %s"
            params.append(status)
        
        query += " ORDER BY r.booking_date DESC LIMIT %s OFFSET %s"
        params.extend([limit, offset])
        
        reservations = execute_query(query, params)
        
        return jsonify({
            'reservations': reservations,
            'pagination': {
                'limit': limit,
                'offset': offset,
                'hasMore': len(reservations) == limit
            }
        })
        
    except Exception as e:
        return jsonify({'message': 'Failed to fetch reservations', 'error': str(e)}), 500

# ===============================================
# NEW ENDPOINTS - ADD THESE TO THE END
# ===============================================

@reservations_bp.route('/owner', methods=['GET'])
@jwt_required()
def get_owner_reservations():
    try:
        claims = get_jwt()
        if claims.get('role') != 'owner':
            return jsonify({'message': 'Owner access required'}), 403
        
        user_id = int(get_jwt_identity())
        status = request.args.get('status')
        limit = int(request.args.get('limit', 50))
        offset = int(request.args.get('offset', 0))
        
        # Build query
        query = """
        SELECT r.*, 
               p.title as property_title, p.address, p.city,
               u.first_name as tenant_first_name, u.last_name as tenant_last_name,
               u.email as tenant_email, u.phone as tenant_phone
        FROM reservations r
        JOIN properties p ON r.property_id = p.property_id
        JOIN users u ON r.tenant_id = u.user_id
        WHERE p.owner_id = %s
        """
        
        params = [user_id]
        
        if status:
            query += " AND r.reservation_status = %s"
            params.append(status)
        
        query += " ORDER BY r.booking_date DESC LIMIT %s OFFSET %s"
        params.extend([limit, offset])
        
        reservations = execute_query(query, params)
        
        return jsonify({
            'reservations': reservations,
            'pagination': {
                'limit': limit,
                'offset': offset,
                'hasMore': len(reservations) == limit
            }
        })
        
    except Exception as e:
        return jsonify({'message': 'Failed to fetch reservations', 'error': str(e)}), 500

@reservations_bp.route('/<int:reservation_id>/status', methods=['PUT'])
@jwt_required()
def update_reservation_status(reservation_id):
    try:
        user_id = int(get_jwt_identity())
        claims = get_jwt()
        role = claims.get('role')
        data = request.get_json()
        
        new_status = data.get('status')
        reason = data.get('reason', '')
        
        if new_status not in ['pending', 'confirmed', 'completed', 'cancelled']:
            return jsonify({'message': 'Invalid status'}), 400
        
        # Get reservation details
        reservation = execute_query(
            """SELECT r.*, p.owner_id 
               FROM reservations r
               JOIN properties p ON r.property_id = p.property_id
               WHERE r.reservation_id = %s""",
            (reservation_id,), fetch_one=True
        )
        
        if not reservation:
            return jsonify({'message': 'Reservation not found'}), 404
        
        # Check permissions
        can_update = False
        if role == 'admin':
            can_update = True
        elif role == 'owner' and reservation['owner_id'] == user_id:
            can_update = True
        elif role == 'tenant' and reservation['tenant_id'] == user_id:
            # Tenants can only cancel their own reservations
            can_update = new_status == 'cancelled'
        
        if not can_update:
            return jsonify({'message': 'Access denied'}), 403
        
        # Update reservation status
        execute_query(
            """UPDATE reservations 
               SET reservation_status = %s, 
                   cancellation_reason = %s
               WHERE reservation_id = %s""",
            (new_status, reason if new_status == 'cancelled' else None, reservation_id)
        )
        
        return jsonify({
            'message': f'Reservation {new_status} successfully',
            'status': new_status
        })
        
    except Exception as e:
        return jsonify({'message': 'Failed to update reservation status', 'error': str(e)}), 500
